src/envoy/http/authn/authn_utils.cc

```diff
@@ -22,51 +22,36 @@ namespace Http {
 namespace Istio {
 namespace AuthN {
 namespace {
-class JsonIterCallback : public Logger::Loggable<Logger::Id::filter> {
- public:
-  // The json iterator callback function
-  static bool JsonIteratorCallback(const std::string&,
-                                   const Envoy::Json::Object&,
-                                   istio::authn::JwtPayload* payload);
-};
+// The JWT audience key name
+static const std::string kJwtAudienceKey = "aud";
 
-bool JsonIterCallback::JsonIteratorCallback(const std::string& key,
-                                            const Envoy::Json::Object& obj,
-                                            istio::authn::JwtPayload* payload) {
-  ENVOY_LOG(debug, "{}: key is {}", __FUNCTION__, key);
-  ::google::protobuf::Map< ::std::string, ::std::string>* claims =
-      payload->mutable_claims();
+// Extract JWT audience into the JwtPayload.
+void ExtractJwtAudience(const Envoy::Json::Object& obj,
+                        istio::authn::JwtPayload* payload) {
+  const std::string& key = kJwtAudienceKey;
+  // "aud" can be either string array or string.
+  // First, try as string, will throw execption if object type is not string.
   try {
-    if (key == "aud") {
-      // "aud" can be either string array or string.
-      // Try as string array, read it as empty array if doesn't exist.
-      std::vector<std::string> aud_vector;
-      try {
-        aud_vector = obj.getStringArray(key, true);
-      } catch (Json::Exception& e) {
-        // Try as string
-        try {
-          auto audience = obj.asString();
-          aud_vector.push_back(audience);
-        } catch (Json::Exception& e) {
-          ENVOY_LOG(error, "aud field type is not string or string array");
-        }
-      }
-      for (size_t i = 0; i < aud_vector.size(); i++) {
-        payload->add_audiences(aud_vector[i]);
-      }
-    } else {
-      // will throw execption if value type is not string.
-      // In current implementation, only string objects are extracted into
-      // claims.
-      (*claims)[key] = obj.asString();
+    std::string obj_str = obj.getString(key);
```

Comment:
I forgot to check one of my previous comment. Why don't you want to check claims instead? try-catch is typically more expensive and we should try to avoid it when possible.

Reply:
Using checking claims["aud"] instead assumes that ExtractJwtAudience() is called after extracting claims. Such calling order assumption tends to be error prone.

Reply:
All calls in this file scope and simple enough to be tested thoroughly (as a matter of fact, doesn't really need a separate function in the first place). I would not sacrifice performance for that.

Reply:
Done.